### PR TITLE
[baseserver] Emit metrics for logs produced by level ENG-349

### PR DESCRIPTION
**Pull request by easyCZ**
## Description

To be able to monitor each cells error logs without having to export them immediately. First we can check volumes. As a side-effect, this also allows us to identify the most offending components in terms of log volume

<details>
<summary>Summary generated by Copilot</summary>

### <samp>🤖 Generated by Copilot at 1a98bf1</samp>

This pull request adds log metrics to the common-go module, which can be used by the base server and other components to report the number of logs by level to prometheus. This enhances the monitoring and debugging capabilities of Gitpod services.

</details>

## Related Issue(s)

Fixes #

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - mp-baseserb73448a74d</li>
	<li><b>🔗 URL</b> - <a href="https://mp-baseserb73448a74d.preview.gitpod-dev.com/workspaces" target="_blank">mp-baseserb73448a74d.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
	<li><b>📦 Version</b> - mp-baseserver-log-metrics-gha.20362</li>
	<li><b>🗒️ Logs</b> - <a href="https://console.cloud.google.com/logs/query;query=jsonPayload.kubernetes.host%3D%22preview-mp-baseserb73448a74d%22%0A%0A--%20Filter%20on%20service:%0A--%20jsonPayload.serviceContext.service%3D%22ws-manager-mk2%22%0A;duration=P1D?project=gitpod-core-dev" target="_blank">GCP Logs Explorer</a></li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

File: components/common-go/log/metrics.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+type LogHook struct {
+	metrics *Metrics
+}
+
+func (h *LogHook) Levels() []logrus.Level {
+	return logrus.AllLevels
+}
+
+func (h *LogHook) Fire(entry *logrus.Entry) error {
+	h.metrics.ReportLog(entry.Level)
```
Comment (mustard-mh) [inline components/common-go/log/metrics.go:60]: log level here is uint32 `type Level uint32`

File: components/gitpod-protocol/src/util/logging.ts [hunk 2]
```diff
@@ -246,7 +247,16 @@ namespace GoogleLogSeverity {
...
 
+const logsCounter = new prometheusClient.Counter({
+    name: "gitpod_logs_total",
+    help: "Total number of logs by level",
+    labelNames: ["level"],
+    registers: [prometheusClient.register],
+});
+
 function doLog(calledViaConsole: boolean, consoleLog: ConsoleLog, severity: GoogleLogSeverity, args: unknown[]): void {
+    logsCounter.labels(severity).inc();
```
Comment (mustard-mh) [inline components/gitpod-protocol/src/util/logging.ts:258]: level here is `type GoogleLogSeverity = "EMERGENCY" | "ALERT" | "CRITICAL" | "ERROR" | "WARNING" | "INFO" | "DEBUG";`

File: components/gitpod-protocol/src/util/logging.ts [hunk 3]
```diff
@@ -246,7 +247,16 @@ namespace GoogleLogSeverity {
     };
 }
 
+const logsCounter = new prometheusClient.Counter({
```
Comment (mustard-mh) [inline components/gitpod-protocol/src/util/logging.ts:250]: Can't find where we use (expose) this counter

**Review by mustard-mh — COMMENTED** [review]
Not all components are using baseserver, do we need to register metric manually for them? Or improve it in follow-up PRs? (search `github.com/gitpod-io/gitpod/common-go/baseserver` in codebase)

Does it make sense to add service name as a label so that we know which components log errors more?

**Comment by easyCZ** [comment]
I've removed TS changes from this PR, I can't figure out how to make it work so will ask EXP

File: components/common-go/log/metrics.go (hunk 1 above)
Comment (easyCZ) [reply components/common-go/log/metrics.go:60]: Yes, but the method calls
```
func (m *Metrics) ReportLog(level logrus.Level) {
	m.logEmitedCounter.WithLabelValues(level.String()).Inc()
}
```
which uses the string version

**Comment by geropl** [comment]
Testing :eyes:

**Review by geropl — APPROVED** [review]
Code LGTM, tested on public-api-server, and works :heavy_check_mark:

**Comment by geropl** [comment]
TS parts are [here](https://github.com/gitpod-io/gitpod/pull/19093); will rebase once this is in.

**Review by vulkoingim — APPROVED** [review]

**Pull request merged**